### DISCUSSION TITLE
Crossplatform db

### DISCUSSION THREAD

**Discussion by ivanjeremic** [oliver-oloughlin/kvdex#156] [Ideas]
Wonder how hard it would be to reimplement this with something like unstorage [https://unstorage.unjs.io/](https://unstorage.unjs.io/) unstorage supports kv stores over many platforms and providers including deno kv.

**Comment by oliver-oloughlin** [comment]
It's an interesting idea. I think there are a few main things that would be needed, if approaching it from a "reuse as much existing code as possible" angle.
1. There would need to be a Deno KV abstraction layer wrapping the unstorage object, in order for kvdex to understand and use the kv client. Something like:
```ts
const storage = createStorage(...)
const kv = kvWrapper(storage)
const db = kvdex(kv, ...)
``` 
2. Deno-specific dependencies, which basically only consists of a few dependencies on Deno std in this case, would need to be replaced.
3. Some types would need to be fixed, as they rely on other types found in the global Deno namespace.
4. Imports would need updating, like removing file extensions and removing npm: prefixes.
5. The Deno Core serializer would need replacing (for example by utilizing Node's v8 serialize instead) or removal.

**Comment by oliver-oloughlin** [comment]
Otherwise I'm sure a similar library could be created specifically targeting unstorage's API instead, but that would likely require a complete re-write from the bottom up, or close to it.
